test/performance/search/index_construction_benchmark.cpp

```diff
@@ -0,0 +1,226 @@
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#include <benchmark/benchmark.h>
+
+#include <seqan3/alphabet/aminoacid/aa27.hpp>
+#include <seqan3/alphabet/nucleotide/dna4.hpp>
+#include <seqan3/range/views/persist.hpp>
+#include <seqan3/range/views/rank_to.hpp>
+#include <seqan3/range/views/to.hpp>
+#include <seqan3/range/views/to_rank.hpp>
+#include <seqan3/search/fm_index/all.hpp>
+#include <seqan3/test/performance/sequence_generator.hpp>
+#include <seqan3/test/seqan2.hpp>
+
+#if SEQAN3_HAS_SEQAN2
+#include <seqan/index.h>
+#endif
+
+static void arguments(benchmark::internal::Benchmark* b)
+{
+    for (int32_t length : {50, 5000, 50'000})
+        b->Args({length, 5});
+
+    b->Args({500, 1'000});
+}
+
+struct fm_index_seqan2;
+struct fm_index_seqan3;
+struct bi_fm_index_seqan2;
+struct bi_fm_index_seqan3;
+
+template <typename index_tag_t, typename rng_t>
+class sequence_generator
+{
+private:
+    size_t length{50'000};
+    size_t collection_size{5};
+
+    using alphabet_t = seqan3::innermost_value_type_t<rng_t>;
+    using inner_rng_t = std::conditional_t<seqan3::dimension_v<rng_t> == 1,
+                                           rng_t,
+                                           std::ranges::range_value_t<rng_t>>;
+    using fm_index_seqan3_t = decltype(seqan3::fm_index{std::declval<rng_t>()});
+    using bi_fm_index_seqan3_t = decltype(seqan3::bi_fm_index{std::declval<rng_t>()});
+
+    inner_rng_t make_single_seqan3(size_t seed = 0)
+    {
+        // The characters with rank 254 and 255 are reserved for the indices.
+        if constexpr (seqan3::alphabet_size<alphabet_t> > 254)
+        {
+            return seqan3::test::generate_sequence<alphabet_t>(length, 0, seed)
+                   | seqan3::views::persist
+                   | seqan3::views::to_rank
+                   | std::views::transform([] (uint8_t const r) { return std::min<uint8_t>(r, 253); })
+                   | seqan3::views::rank_to<alphabet_t>
+                   | seqan3::views::to<inner_rng_t>;
+        }
+        else
+        {
+            return seqan3::test::generate_sequence<alphabet_t>(length, 0, seed) | seqan3::views::to<inner_rng_t>;
+        }
+    }
+
+    rng_t make_collection_seqan3()
+    {
+        rng_t result;
+        for (size_t i = 0; i < collection_size; ++i)
+            result.emplace_back(make_single_seqan3(i));
+
+        return result;
+    }
+
+    rng_t make_sequence_seqan3()
+    {
+        if constexpr(seqan3::dimension_v<rng_t> == 1)
+            return make_single_seqan3();
+        else
+            return make_collection_seqan3();
+    }
+
+#if SEQAN3_HAS_SEQAN2
+    // Map the SeqAn3 alphabet to its SeqAn2 equivalent.
+    using seqan2_alphabet_t = std::conditional_t<std::same_as<alphabet_t, seqan3::dna4>, seqan::Dna,
```

Comment:
why do you need to do this here? 
Can't you write directly into the benchmark template:

```diff
BENCHMARK_TEMPLATE(index_benchmark, bi_fm_index_seqan3, std::vector<seqan3::dna4>)->Apply(arguments);
#if SEQAN3_HAS_SEQAN2
- BENCHMARK_TEMPLATE(index_benchmark, bi_fm_index_seqan2, std::vector<seqan3::dna4>)->Apply(arguments);
+ BENCHMARK_TEMPLATE(index_benchmark, bi_fm_index_seqan2, seqan::String<seqan::Dna>)->Apply(arguments);
#endif // SEQAN3_HAS_SEQAN2
```

This also seems error-prone if change the alphabet to `dna5` for testing, this function will fall back to `char` for seqan2 wihtout any error